### PR TITLE
feat(dart/transform): Detect annotations which extend Injectable or Template.

### PR DESCRIPTION
**Pull request by jakemac53**
Create a method that recursively walks imports from an entry point and
determines where classes are registered.

Use this information to determine if a particular annotation implements or
extends Injectable or Template.

### CODE REVIEW

File: modules/angular2/src/transform/common/classdef_parser.dart [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+import 'dart:async';
+
+import 'package:analyzer/analyzer.dart';
+import 'package:angular2/src/transform/common/asset_reader.dart';
+import 'package:angular2/src/transform/common/logging.dart';
+import 'package:barback/barback.dart';
+import 'package:code_transformers/assets.dart';
+
+Future<Map<AssetId, List<ClassDeclaration>>> createTypeMap(
```
Comment (jakemac53) [inline modules/angular2/src/transform/common/classdef_parser.dart]: Fyi, this returns a Map to support future some future use cases. @kegluneq would have more info on this (its actually cherry-picked with a few tweaks from his branch).

Comment (yjbanov) [inline modules/angular2/src/transform/common/classdef_parser.dart]: I know we haven't been good at this, but it's never too late to start. Let's write docs for public functions.

File: modules/angular2/src/transform/common/classdef_parser.dart [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    AssetReader reader, AssetId id) {
+  return _recurse(reader, id);
+}
+
+Future<Map<AssetId, List<ClassDeclaration>>> _recurse(
+    AssetReader reader, AssetId id, [visitor, seen]) async {
```
Comment (yjbanov) [inline modules/angular2/src/transform/common/classdef_parser.dart]: Why no types on visitor and seen?

File: modules/angular2/src/transform/template_compiler/classdef_parser.dart [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import 'package:angular2/src/dom/dom_adapter.dart';
+import 'package:angular2/src/reflection/reflection.dart';
+import 'package:angular2/src/transform/common/asset_reader.dart';
+import 'package:angular2/src/transform/common/logging.dart';
+import 'package:angular2/src/transform/common/names.dart';
+import 'package:angular2/src/transform/common/parser.dart';
+import 'package:barback/barback.dart';
+import 'package:code_transformers/assets.dart';
+
+Future<Map<AssetId, List<ClassDeclaration>>> createTypeMap(
```
Comment (yjbanov) [inline modules/angular2/src/transform/template_compiler/classdef_parser.dart]: Is this sufficiently different from the above to warrant a separate implementation?

Comment (jakemac53) [inline modules/angular2/src/transform/template_compiler/classdef_parser.dart]: Whoops this just slipped in. This was the old location and I moved it to the common folder.

File: modules/angular2/src/transform/common/classdef_parser.dart (hunk 2 above)
Comment (jakemac53) [inline modules/angular2/src/transform/common/classdef_parser.dart]: They are only used for recursive calls to this function so they didn't provide a lot of value, but I have added them anyways.

File: modules/angular2/src/transform/common/classdef_parser.dart (hunk 1 above)
Comment (jakemac53) [inline modules/angular2/src/transform/common/classdef_parser.dart]: done

**Comment by jakemac53** [comment]
ok I pushed a new commit so you can see the changes, I will squash before submitting though

**Comment by jakemac53** [comment]
Also, I noticed the Travis bots are running out of memory when doing pub build, is that a known issue or do you think its a result of this cl?

**Comment by yjbanov** [comment]
Could you rebase and try running travis again? I haven't seen the build run out of memory on Travis, although webdriver's java process runs out of memory on my laptop quite frequently.

**Comment by yjbanov** [comment]
LGTM if LGT Travis

**Comment by jakemac53** [comment]
need to clean up the history here sorry, had to scrap my whole directory due to some weird gulp problems that I couldn't resolve :(

**Comment by jakemac53** [comment]
Hmm, looks like pub get from travis is unable to pick up the new version of the analyzer for some reason? When I run pub get locally it works fine. Maybe I will just give this a few hours and try again...

**Pull request merged**